File: worker/src/oidc.ts
/**
 * Verify that a request really came from this repository's Actions run.
 *
 * A port of bridge/src/bridge/oidc.py. The write back workflow holds no Figma credential and no
 * shared secret. It asks GitHub for a short lived token saying which repository, workflow and
 * event produced it, and this checks that token against GitHub's published keys.
 *
 * What that buys: this Worker can be a public address. Anything unsigned is refused, a token from
 * somebody else's repository is refused, and there is no long lived credential in a repository
 * secret for every current and future workflow to read.
 */
import { createRemoteJWKSet, jwtVerify } from 'jose'

export const ISSUER = 'https://token.actions.githubusercontent.com'
export const AUDIENCE = 'figma-bridge'

export class NotFromActions extends Error {}

export interface Verifier {
  verify(authorization: string | null): Promise<Record<string, unknown>>
}

export class ActionsVerifier implements Verifier {
  private readonly jwks: ReturnType<typeof createRemoteJWKSet>

  constructor(private readonly repo: string) {
    // createRemoteJWKSet caches the key set per isolate, so a warm Worker does not refetch it on
    // every request and a cold one pays a single round trip.
    this.jwks = createRemoteJWKSet(new URL(`${ISSUER}/.well-known/jwks`))
  }

  async verify(authorization: string | null): Promise<Record<string, unknown>> {
    if (!authorization?.toLowerCase().startsWith('bearer ')) {
      throw new NotFromActions('no bearer token')
    }
    const token = authorization.slice(7).trim()

    let claims: Record<string, unknown>
    try {
      const result = await jwtVerify(token, this.jwks, {
        issuer: ISSUER,
        audience: AUDIENCE,
        algorithms: ['RS256'],
        requiredClaims: ['exp', 'iat', 'iss', 'aud', 'sub'],
      })
      claims = result.payload as Record<string, unknown>
    } catch (error) {
      throw new NotFromActions(`token did not verify: ${(error as Error).message}`)
    }

    if (claims.repository !== this.repo) {
      // Not worth explaining to the caller: it would confirm what this guards to whoever is
      // probing it.
      throw new NotFromActions('token is for another repository')
    }
    if (typeof claims.iat === 'number' && claims.iat > Date.now() / 1000 + 60) {
      throw new NotFromActions('token was issued in the future')
    }
    return claims
  }
}

File: bridge/src/bridge/oidc.py
"""Verify that a request really came from this repository's Actions run.

The write back workflow holds no Figma credential and no shared secret. It asks GitHub for a
short lived token that says which repository, workflow and event produced it, and this module
checks that token against GitHub's published keys.

What that buys: the bridge can be a public address. Anything unsigned is refused, a token from
somebody else's repository is refused, and there is no long lived credential sitting in a
repository secret for every current and future workflow to read.
"""

from __future__ import annotations

import time

import httpx
import jwt
from jwt import PyJWKClient

ISSUER = "https://token.actions.githubusercontent.com"
AUDIENCE = "figma-bridge"


class NotFromActions(Exception):
    """The token is missing, malformed, expired, or from somewhere else."""


class ActionsVerifier:
    def __init__(self, repo: str, jwks_client: PyJWKClient | None = None) -> None:
        self._repo = repo
        self._jwks = jwks_client or PyJWKClient(f"{ISSUER}/.well-known/jwks", lifespan=3600)

    def verify(self, authorization: str | None) -> dict:
        if not authorization or not authorization.lower().startswith("bearer "):
            raise NotFromActions("no bearer token")
        token = authorization.split(" ", 1)[1].strip()
        try:
            key = self._jwks.get_signing_key_from_jwt(token).key
            claims = jwt.decode(
                token,
                key,
                algorithms=["RS256"],
                audience=AUDIENCE,
                issuer=ISSUER,
                options={"require": ["exp", "iat", "iss", "aud", "sub"]},
            )
        except (jwt.PyJWTError, httpx.HTTPError) as error:
            raise NotFromActions(f"token did not verify: {error}") from error

        if claims.get("repository") != self._repo:
            # Not an error worth explaining to the caller: it would confirm what the bridge
            # guards to whoever is probing it.
            raise NotFromActions("token is for another repository")
        if claims.get("iat", 0) > time.time() + 60:
            raise NotFromActions("token was issued in the future")
        return claims
